src/object/schema_manager.c

```diff
@@ -15641,21 +15645,58 @@ sm_collect_truncatable_classes (MOP class_mop, std::unordered_set < OID > &trun_
       return er_errid ();
     }
 
-  trun_classes.emplace (*ws_oid (class_mop));
-
   pk_constraint = classobj_find_cons_primary_key (class_->constraints);
-  if (pk_constraint == NULL || pk_constraint->fk_info == NULL)
+  is_pk_referred = pk_constraint != NULL && pk_constraint->fk_info != NULL;
+
+  if (is_pk_referred && !is_cascade)
     {
-      /* if no PK or FK-referred, it can be truncated */
-      return NO_ERROR;
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_TRUNCATE_PK_REFERRED, 1, pk_constraint->fk_info->name);
+      return ER_TRUNCATE_PK_REFERRED;
     }
 
-  /* Now, there is a PK, and are some FKs-referring to the PK */
+  error = sm_partitioned_class_type (class_mop, &partition_type, NULL, NULL);
+  if (error != NO_ERROR)
+    {
+      return error;
+    }
 
-  if (!is_cascade)
+  switch (partition_type)
     {
-      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_TRUNCATE_PK_REFERRED, 1, pk_constraint->fk_info->name);
-      return ER_TRUNCATE_PK_REFERRED;
+    case DB_PARTITION_CLASS:
+      /*
+       * TRUNCATE [partition class] is already prevented.
+       * This can be reached because if a FK of partitioned class referes a PK, FKs of the partition classes also refer to it.
+       *
+       * No need to be added to trun_classes:
+       *  Partition classes will be added when examining the partitioned class of them.
+       * No need to check FK for cascaing:
+       *  It's enough to check the PK of the partitioned class for cascading.
+       */
+      return NO_ERROR;
+    case DB_PARTITIONED_CLASS:
+      /* Find partition classes if exists */
+      assert (class_->users);
+      for (subs = class_->users; subs; subs = subs->next)
+	{
+	  error = au_fetch_class (subs->op, &subclass, AU_FETCH_READ, DB_AUTH_ALTER);
+	  if (error != NO_ERROR)
+	    {
+	      return error;
+	    }
+	  if (subclass->partition)
```

Comment:
Is the value of 'subclass->partition' always not null in this case? If it is, how about use the assert function?

Reply:
Yes, you're right. It is a pre-condition rather than an if-condition. I'll change it to assertion.

Reply:
I mimicked some code in `do_drop_partitioned_class()`, and it seems to assume that partitioned class can have child classes which are not partition class. However, it is impossible to inherit a partitioned table. Assuming it is a legacy code, I've decided not to consider that case.

Reply:
Or we can be conservative that just leave it like before with the assumption that another feature but partitioning use inheritance hierarchy.